File: 251003/Lysiakoff/rvlab5/rvlab4_publisher/src/main/java/com/publisher/rvlab4_publisher/service/MarkService.java
package com.publisher.rvlab4_publisher.service;

import com.publisher.rvlab4_publisher.dto.MarkRequestTo;
import com.publisher.rvlab4_publisher.dto.MarkResponseTo;
import com.publisher.rvlab4_publisher.entity.Mark;
import com.publisher.rvlab4_publisher.repository.MarkRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class MarkService {
    private final MarkRepository markRepository;

    public List<MarkResponseTo> getAllMarks() {
        return markRepository.findAll().stream()
                .map(mark -> new MarkResponseTo(mark.getId(), mark.getName()))
                .collect(Collectors.toList());
    }

    public MarkResponseTo getMarkById(Long id) {
        return markRepository.findById(id)
                .map(mark -> new MarkResponseTo(mark.getId(), mark.getName()))
                .orElseThrow(() -> new NoSuchElementException("Mark not found"));
    }

    public Optional<Mark> getMarkByName(String name) {
        return markRepository.findByName(name);
    }

    public MarkResponseTo createMark(MarkRequestTo request) {
        Mark mark = new Mark();
        mark.setName(request.getName());

        Mark savedMark = markRepository.save(mark);

        return new MarkResponseTo(
                savedMark.getId(),
                savedMark.getName()
        );
    }

    public MarkResponseTo updateMark(Long id, MarkRequestTo request) {
        Mark mark = markRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Mark not found"));
        mark.setName(request.getName());
        markRepository.save(mark);
        return new MarkResponseTo(id, mark.getName());
    }

    public void deleteMark(Long id) {
        Mark mark = markRepository.findById(id).orElseThrow(() -> new NoSuchElementException("Mark no found"));
        markRepository.deleteById(id);
    }
}

File: 251003/Kukhatskavolets/microblog/src/main/kotlin/by/kukhatskavolets/services/MarkService.kt
package by.kukhatskavolets.services

import by.kukhatskavolets.dto.requests.MarkRequestTo
import by.kukhatskavolets.dto.responses.MarkResponseTo
import by.kukhatskavolets.mappers.toEntity
import by.kukhatskavolets.mappers.toResponse
import by.kukhatskavolets.repositories.MarkRepository
import by.kukhatskavolets.repositories.TweetRepository
import org.springframework.stereotype.Service

@Service
class MarkService(
    private val markRepository: MarkRepository,
    private val tweetRepository: TweetRepository,
) {
    fun createMark(markRequestTo: MarkRequestTo): MarkResponseTo {
        val mark = markRequestTo.toEntity()
        val savedMark = markRepository.save(mark)
        return savedMark.toResponse()
    }

    fun getMarkById(id: Long): MarkResponseTo {
        val mark = markRepository.findById(id).orElseThrow { NoSuchElementException() }
        return mark.toResponse()
    }

    fun getAllMarks(): List<MarkResponseTo> =
        markRepository.findAll().map { it.toResponse() }

    fun updateMark(id: Long, markRequestTo: MarkRequestTo): MarkResponseTo {
        val updatedMark = markRequestTo.toEntity().apply { this.id = id }
        return markRepository.save(updatedMark).toResponse()
    }

    fun deleteMark(id: Long) {
        markRepository.findById(id).orElseThrow { NoSuchElementException() }
        markRepository.deleteById(id)
    }

    fun getMarksByTweetId(tweetId: Long): List<MarkResponseTo> {
        val tweet = tweetRepository.findById(tweetId).orElseThrow { NoSuchElementException() }
        return tweet.marks.map { it.toResponse() }
    }
}